Add unread product reviews filter

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->

Add unread product reviews filter
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
As part of the Hack Week project, I added a new filter for showing only `unread` product reviews. This comes from some user feedback stating things like: 

> Hi tech team. I suggest you change the reviews, so it’s possible to see which reviews the user hasn’t read. I have 2 unread reviews and I have opened all current reviews and now I need to scroll to the bottom and reload older reviews. Also when returning from a specific review, return to where you were in the list of reviews. Thank you.

How filtering unread product reviews work: 
- The `unread` state of product reviews relies on WPcom notifications `unread` status. 
- WC Rest API doesn't provide a way to fetch unread product reviews, as it doesn't keep track of this status. So we have to deal with this locally. 
- In the apps what we do is retrieve all the `unread` product review notifications from DB. Then we use `meta?.ids?.comment` field from the notification model to get the product review id. 
- With the list of unread product review ids we can now fetch from the API those specific reviews by setting `productIds: List<Long>?` field inside the `FetchProductReviewsPayload` 
- The rest of the code changes are essentially to handle pagination properly for unread product reviews. The unread reviews will be loaded in chunks of 25 items until all unread reviews are fetched. 

Additionally, this PR gets rid of some code duplication and removes `https://github.com/woocommerce/woocommerce-android/pull/9753/files#diff-67759c42aa62e65e7e1f76c809cd9804580b532574f8ee909481f2db6cecf0c2` from the code and refactors `ReviewListRepository` to support fetching product reviews for a specific product.  


### To Improve
When caching product reviews there are some scenarios when reopening product reviews list will display an outdated list for a couple seconds (until the new data is loaded). This is specially noticeable when you:
- Open product reviews
- Apply unread filter
- Close product reviews
- Reopen it and see how only unread reviews are shown for a couple seconds. 


### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
1. Open a store that has a considerable amount of product reviews. You can [use this plugin to generate fake product reviews](https://wordpress.org/plugins/woo-virtual-reviews/). Which by the way I don't know how that is even allowed. Pretty sketchy 😅 
2. Mark several reviews as read
3. Enable and disable the unread filter and check the list is updated correctly. 

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

**Unread reviews filter demo**

https://github.com/woocommerce/woocommerce-android/assets/2663464/82965416-c1ab-4487-8078-925030ce93a4

**Empty unread reviews UI**

****

https://github.com/woocommerce/woocommerce-android/assets/2663464/cb5ac5b6-dc0f-4be5-a88d-b2ff5de86fc0




- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

JorgeMucientes committed:
Minor code optimization

JorgeMucientes committed:
Add unread reviews filter widget to ReviewListFragment

JorgeMucientes committed:
Filter product reviews from more menu by "is unread" status

JorgeMucientes committed:
Load more when filtering by "unread" state shows only a few reviews

JorgeMucientes committed:
UI improvements over unread filter

JorgeMucientes committed:
Automatically load more reviews when unread reviews empty

JorgeMucientes committed:
Automatically load more unread reviews when applying unread filter

JorgeMucientes committed:
Improve empty case message

JorgeMucientes committed:
Fix detekt issue

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListRepository.kt:
```diff
@@ -168,15 +168,14 @@ class ReviewListRepository @Inject constructor(
-                productsMap.containsKey(it.remoteProductId) && productsMap[it.remoteProductId] != null
-            }.also { review ->
-                review.forEach {
+            cachedReviews = cachedReviews
+                .filter {
+                    // Only returns reviews that have a matching product in the db.
+                    productsMap.containsKey(it.remoteProductId) && productsMap[it.remoteProductId] != null
+                }.onEach {
```
Just a minor code simplification

JorgeMucientes committed:
Implement unread reviews filtering in product detail review list

JorgeMucientes committed:
Show unread product reviews for product detail review list

JorgeMucientes committed:
Remove unneeded empty code blocks

JorgeMucientes committed:
Fix bug, displaying read reviews when unread filter is enabled

JorgeMucientes committed:
Handle empty case for unread reviews in ProductReviewsFragment

JorgeMucientes committed:
Merge branch 'trunk' into hack-week/unread-product-reviews-filter

JorgeMucientes committed:
Updates release notes

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/reviews/ProductReviewsRepository.kt:
```diff
@@ -58,9 +63,26 @@ class ProductReviewsRepository @Inject constructor(
                 localSiteId = selectedSite.get().id,
                 remoteProductId = remoteProductId
             ).map { it.toAppModel() }
+            val reviewRemoteIdToReadStatusMap = getReviewNotificationReadValueByRemoteId()
+            return@withContext reviews.onEach {
+                it.read = reviewRemoteIdToReadStatusMap[it.remoteId] ?: false
+            }
         }
+
+    /**
+     * Uses the product review notifications to create a map of
+     * [org.wordpress.android.fluxc.model.notification.NotificationModel.read] by [ProductReview.remoteId].
+     */
+    private suspend fun getReviewNotificationReadValueByRemoteId(): Map<Long, Boolean> {
+        return withContext(Dispatchers.IO) {
+            notificationStore.getNotificationsForSite(
+                site = selectedSite.get(),
+                filterBySubtype = listOf(STORE_REVIEW.toString())
+            ).associate { it.getCommentId() to it.read }
+        }
+    }
```
The product reviews displayed from ProductDetail did not support showing the `unread` status. So I've added this code to support that as well as the `unread/read` filter.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/reviews/ProductReviewsRepository.kt:
```diff
@@ -58,9 +63,26 @@ class ProductReviewsRepository @Inject constructor(
+     * Uses the product review notifications to create a map of
+     * [org.wordpress.android.fluxc.model.notification.NotificationModel.read] by [ProductReview.remoteId].
+     */
+    private suspend fun getReviewNotificationReadValueByRemoteId(): Map<Long, Boolean> {
+        return withContext(Dispatchers.IO) {
+            notificationStore.getNotificationsForSite(
+                site = selectedSite.get(),
+                filterBySubtype = listOf(STORE_REVIEW.toString())
+            ).associate { it.getCommentId() to it.read }
+        }
+    }
```
I know this is how it was already done in the reviews list, but I think this is something we should think about improving (not necessarily in this PR): here we get the list of all notifications from the DB then do the filtering/checking on code side, this might cause issues for large stores.
What we should do instead is: add a new function to the `NotificationStore` that allows retrieving notifications matching a certain `commentId`, this way we'll retrieve only the values we are interested in.

This won't be simple to implement, since the `commentId` is not a column, but a value nested in a Json object, but I think it's still doable.

JorgeMucientes replied:
Hmm, I don't have a strong opinion about adopting that approach. But the refactor is considerable and I think it's hardly justifiable to the benefits it provides. When you mention: 
> this might cause issues for large stores.

I presume you are talking about performance issues? However, since `commentId` is not a column, you'd still need to retrieve all the notifications from the db, parse them from json and then filter them in code. So, I'm unsure what the benefits are here.

hichamboushaba replied:
Just for context, I mentioned this to keep track of it, I didn't mean it should be implemented now.

> However, since commentId is not a column, you'd still need to retrieve all the notifications from the db, parse them from json and then filter them in code. So, I'm unsure what the benefits are here.

I know the `commentId` is stored in in a String (json representation), so I was thinking about two approaches:
1. Either figure out a way to play with Sql's String comparison operators to have a working condition there, it could be as simple as: multiple expressions of `LIKE '%{id}%` combined with an `OR` operator.
2. If the above turns out unreliable, we can then think about extracting the `commentId` to its own column.

Also, when mentioning performance issues, I'm thinking about medium to large stores, which would have thousands of reviews/notifications, if we fetch the whole list to the app's memory, we are taking a lot of risk.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListRepository.kt:
```diff
@@ -43,6 +43,7 @@ class ReviewListRepository @Inject constructor(
 ) {
     companion object {
         private const val PAGE_SIZE = WCProductStore.NUM_REVIEWS_PER_FETCH
+        const val MIN_NUMBER_OF_ITEMS_TO_MAKE_CONTENT_SCROLL = 10
```
For future readers, this variable will need some clarification, why it's needed, and why it does.

JorgeMucientes replied:
I've changed that, to properly support bigger screens. Additionally I've added a comment explaining what that value is. Changes done here: https://github.com/woocommerce/woocommerce-android/pull/9753/commits/15a48ce063cd154304f9913b37ddab21a3a237d4

hichamboushaba reviewed (commented):
Thanks @JorgeMucientes for working on this 💯, nice work in trying to find a workaround for the API limitation.
Still, I'm not sure this is ready for user release, as you can see from the following video:

https://github.com/woocommerce/woocommerce-android/assets/1657201/e38c9d5a-df9a-43d1-a82f-bb494973e87d

The feature has two major issues:
1. The loading can sometimes be very long, and this is because we might end-up loading multiple pages before reaching the first unread review, and depending on the site speed, and the number of reviews, the users might just quit thinking that it doesn't work.
2. Because of this [check](https://github.com/woocommerce/woocommerce-android/blob/27c744fa93268b373408a4a137249e91eff60e37/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt#L183), The app loads a limited number of unread review (10 or just a little bit over it), and this might be misleading, for example for the video above, I have 119 unread reviews, but with the filter, I see only 13.

I think this should be discussed with product to explain the API limitations, and decide whether we are OK with it.

JorgeMucientes committed:
Fix bug not loading all unread reviews

JorgeMucientes committed:
Add different constant value for phones and big screens (tablets)

JorgeMucientes commented:
Hey @hichamboushaba thank you for the detailed review. 
About the major issues you mentioned: 

> The loading can sometimes be very long, and this is because we might end-up loading multiple pages before reaching the first unread review, and depending on the site speed, and the number of reviews,

I don't think there's much we can do here. As you well noticed, the API doesn't provide a way to filter reviews by `read/unread` status and this is a "best effort" workaround solution. I think a long loading time situation should not be a common scenario, and if it happens is a compromise we have to assume given the current restrictions. Moreover, if the user hosting/site is slow, many other areas of the app will be slow. This is a problem we've been “suffering” within Woo app from the start, of whom many users have complained about. But there's little we can do and that didn't prevent us from releasing features.

Btw, it seems from your screen recording that you triggered enabling the unread filter and swipe to refresh at the same time? That might also impact the speed of the app to display results. 

> Because of this [check](https://github.com/woocommerce/woocommerce-android/blob/27c744fa93268b373408a4a137249e91eff60e37/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt#L183), The app loads a limited number of unread review (10 or just a little bit over it), and this might be misleading, for example for the video above, I have 119 unread reviews, but with the filter, I see only 13.

Hmm, good catch! That's not how the algorithm was intended to work 😅. The idea was that when you enable the `unread` filter, the code ensures that at least 10 unread items are displayed on the screen. Why 10? Because typically, a phone screen will fit 5-7 items at once. Displaying 10 will make the users scroll and trigger the loading of the next page. But the problem is (as you noticed) that if in the next 25 items there are no unread reviews, and you are already at the end of the list, then no more pages will be loaded. 
I think that problem can be fixed by adding one extra check. I've pushed the changes here https://github.com/woocommerce/woocommerce-android/pull/9753/commits/aa09708cdd9787dc8ef8229c7fb87a426c48c83c. This should fix the described problem.

hichamboushaba commented:
> I don't think there's much we can do here

I agree 🙂, but the question we should be asking ourselves: is this too important and urgent? and if yes, can't we do something on the backend side to support it?

> I think a long loading time situation should not be a common scenario

I respectfully disagree, let's just imaging a simple scenario: a medium store that has 1000 review (this is still a small number), and where the user have already marked the 900 most recent reviews as read (something that could happen when, the user start following reviews closely just later, and the first ones stay `unread`), in this case, if the user opens the reviews screen, then toggles the `unread` filter, the app will perform 36 requests before getting the first unread review, and 36 requests will be very long even for fast servers.
In the past, we had a similar issue with a similar limitation on the API: SKU search, the API didn't support it, and a way to support was to do something like this PR by filtering locally, I wasn't involved in the discussions, but I know the team didn't proceed with the implementation, until the API started supporting it.

> Hmm, good catch! That's not how the algorithm was intended to work 😅. The idea was that when you enable the unread filter, the code ensures that at least 10 unread items are displayed on the screen

Huh, I misunderstood your intention then, I thought you added this as a way to prevent a large number of requests, and I thought it might be a good compromise 😄, I'll explain why: without this, if the user have just new unread reviews, but all of the past one were already read, then when they'll load the screen, and toggle the filter, when they scroll a bit, the app will keep fetching the server's pages until they are exhausted, and depending on the size of the store, this will be a very big number of requests.

---
Just as a final note, my goal is not to block the PR, I'm just putting myself on the user's feet, and also it's not my call whether to accept the user experience of the PR or not, we just need to get the product's opinion about it 🙂.

JorgeMucientes committed:
Fix detekt indentation issues

JorgeMucientes commented:
Hey @pmusolino 👋🏼 We need your input here. There's some discussion about the UX compromises we need to assume if we merge the unread filter feature as it is now or if is better to pause the merging and invest more time to provide a better experience. We'd like know what your thoughts are. 

Let me first provide some context and summarize the compromises:

- The current API for fetching product reviews, doesn't provide the status of whether the review is `read/unread`. This status is calculated locally by matching product reviews with WP.com notifications. For this reason, applying the `unread` filter is also something that is calculated locally.
- Without getting into much tech details. The current way of filtering `unread` reviews has the following tradeoffs: 
  - **Very long loading times** when applying the filter in the following scenario: With hundreds of product reviews and the unread ones are the oldest ones (at the end of the list). 
  - **Many API request triggered** in the same scenario mentioned above: when the unread reviews are at the end of the list and there are hundreds of them. 

There are ways to mitigate this that would involve extra work. Although I'm not sure how to estimate that extra effort, because part of it would require some back end changes. We might be able to tackle those back end changes ourselves, but I wouldn't dare to estimate how much effort that would involve because I honestly have no idea. 

Imho the compromises are acceptable to merge this feature as is. I believe we can release it, measure the feature usage and then decide if its worth investing extra effort to mitigate the long loading times and the server load. I think @hichamboushaba disagrees to some extent, and is more in favor of investing extra effort before merging to provide a better experience from the beginning. 

Let us know what your though are 🙂

Also @hichamboushaba feel free to add any extra arguments here that I might have missed. And once again thanks for the careful review and triggering the discussion 🙂.

JorgeMucientes committed:
Merge branch 'trunk' into hack-week/unread-product-reviews-filter

JorgeMucientes committed:
Fix release notes after wrong merge

pmusolino commented:
Thanks for involving me in the discussion @JorgeMucientes!

> Many API request triggered in the same scenario mentioned above: when the unread reviews are at the end of the list and there are hundreds of them.

If we have for example 100 pages of reviews, how do you know that the last unread review is for example at page 10? I think you need, in any case (also if the review is at page 1), do download all the reviews at the same time to have them stored locally, before making the query. Am I right?
If am I right, I think this could be an issue, as suggested by Hicham, because it means we are always going to download tons of reviews for this.

However, I also understand the need to measure the feature usage. So, my suggestion would be, if possible and if it doesn't require much effort, to apply the filter only on the last 200-300 reviews -based also on how much reviews we can fetch for each page- and on the one already stored locally, just to understand how many users are going to use this filter. Based on the feedback, we can decide whether to invest further in this feature or not, maybe improving the feature adding a way in the core API to filter them. My proposal it's also not ideal, and before proceeding with it, I would suggest to do the thing which makes more sense: did you asked the core team if it is possible to prioritize this implementation on their side?

JorgeMucientes commented:
Thanks for the input @pmusolino. 

> I think you need, in any case (also if the review is at page 1), do download all the reviews at the same time to have them stored locally, before making the query. Am I right?

About that, not exactly. There's no need to download all the reviews to always load the last unread review. We'll fetch reviews from the API and filter them as needed. Let me put an example. 
- User enters product reviews list for the first time. 25 items are fetched from API (25 is page size)
- User enables the unread only filter. Let's say the user has 1000 reviews in total and the unread reviews happen to be from 100 to 115 and from 900 to 950 positions. 
- The app will start fetching from API in blocks of 25 items until we reach the first unread reviews. After 4 requests we'll reach position 125. Since we have unread reviews from position 100 to 115 those will be displayed. No more requests will be triggered. 
- Now users scrolls down to the last of the unread reviews displayed, pagination will be triggered again and this time 35 requests will be fired until we reach the next chunk of unread reviews at position 900. This situation is indeed a problem. Although would only happen when the next `unread` reviews are very far away in from each other. 

This should only happen once because whenever reviews are fetched from the API, they are saved in the DB. However is not the case, we always refresh reviews from the API when entering the screen. Probably to keep the other review statuses updated: approved, hold, spam, unspam, trash and untrash. 

> if possible and if it doesn't require much effort, to apply the filter only on the last 200-300 reviews -based also on how much reviews we can fetch for each page

Yeah that'd be possible. We could fetch the next 3-4 pages and if there's no unread review then stop fetching there. Although that might be seen as a bug if users are aware that they have more `unread` reviews and we're not loading them or telling them that they don't have any unread review. Specially if we show the bardge count for unread reviews 😅.

> did you asked the core team if it is possible to prioritize this implementation on their side?

~~No, I haven't. I'm not sure who to ask though. Any suggestion? ~~ NO NEED: see edit below

About possible backend solutions just keep this in mind (quoting Hicham): 

> it’s impossible for them to support it, because notifications are part of WPCom unless I’m mistaken.
But still, if we invest some time, we can come with an acceptable solution, just a small change to WC’s API to allow returning a list of reviews by their IDs, it should be very simple to implement, and one of us could do it, if this is supported, then we can invert the algorithm:
> - Get the ids of the undread reviews (using what we have in the app now)
> - Fetch reviews using those ids

EDIT: Actually, the suggested implementation by @hichamboushaba  might not require any backend changes after all. I've just learned that [WC Rest API support fetching product reviews by specific ids. ](https://woocommerce.github.io/woocommerce-rest-api-docs/#retrieve-a-product-review:~:text=%5D%0A%20%20%20%20%7D%0A%7D-,List%20all%20product%20reviews,-This%20API%20lets). Following that approach **will fix all the compromises but requires a considerable refactor** to the current implementation. I'm happy to work on it but could be roughly 3 days work. Are we ok with this @pmusolino ?

JorgeMucientes commented:
Moving this PR to draft. I'll be tackling the suggested solution and reopen the PR when ready. Thanks for the feedback folks. 

cc @hichamboushaba

JorgeMucientes committed:
Retrieve product reviews by id

JorgeMucientes committed:
Merge branch 'trunk' into hack-week/unread-product-reviews-filter

JorgeMucientes committed:
Handle pagination for unread reviews

JorgeMucientes committed:
Fix UI loading states while switching between unread/read filter

JorgeMucientes committed:
Fix pagination issues for unread reviews

JorgeMucientes committed:
Fix crash for empty case

JorgeMucientes committed:
Cancel previously launched review fetching jobs to avoid UI overlappings

JorgeMucientes committed:
Remove unused code from old implementation

JorgeMucientes committed:
Improve readability

JorgeMucientes committed:
Add error handling and improve code readability

JorgeMucientes committed:
Remove unneeded coroutine launch

JorgeMucientes committed:
Make explicit when we are deleting the existing cached reviews

JorgeMucientes committed:
Fix on scroll pagination issues

Avoid cancelling fetch tasks when scrolling

JorgeMucientes committed:
Update gradle FluxC changeset

JorgeMucientes committed:
Update release notes

JorgeMucientes committed:
Fix loading state when disabling filter

JorgeMucientes committed:
Add unread filtering to product detail -> product reviews

JorgeMucientes committed:
Update tracking for failed requests

JorgeMucientes committed:
Remove repository with duplicated functionality

ProductReviewsRepository.kt can be removed as ReviewListRepository already provided the needed logic to handle fetching product reviews.

JorgeMucientes committed:
Fix detekt indentation issue

JorgeMucientes committed:
Fix unit tests

JorgeMucientes committed:
Add couple unit tests for new unread filtering

JorgeMucientes committed:
Removes unnecessary view state update

JorgeMucientes committed:
Avoid showing cached reviews that don't belong to the current product

JorgeMucientes committed:
Load all unread reviews and then filter by product id if needed

Since it is not possible to filter unread reviews notifications by product id, we are forced to fetch unread reviews for all products and then filer locally.

JorgeMucientes commented:
Hey @hichamboushaba I've updated the product description and pushed all the changes of a revamped implementation for filtering unread product reviews. The PR is ready for another round 🙏🏼

hichamboushaba commented on RELEASE-NOTES.txt:
```diff
@@ -1,9 +1,17 @@
+- [**] [Internal] In-Person Payments: if WcPay setup not finished then we allow a user to finish it from the app [https://github.com/woocommerce/woocommerce-android/pull/9732]
+- [*] [Internal] Store Onboarding: show an explanation screen before open the Payments task in a WebView [https://github.com/woocommerce/woocommerce-android/pull/9772]
+- [**] Order creation: Merchants can now select a tax rate when creating an order [https://github.com/woocommerce/woocommerce-android/pull/9773]
+- [**] Product editing: Merchants can now add downloadable files from all supported file types [https://github.com/woocommerce/woocommerce-android/pull/9745]
```
Before merging this, please make sure you check this, it seems weird that we are "adding" new entries for 15.4, merging trunk would probably fix this, but just keep it in mind 🙂.

JorgeMucientes replied:
Yeah! this should be fixed once I update the branch with `trunk`. I'll deal with it as soon as we are done with changes on the PR and I've added the correct FluxC changeset.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/reviews/ProductReviewsViewModel.kt:
```diff
@@ -87,101 +96,118 @@ class ProductReviewsViewModel @Inject constructor(
+        launch {
+            // Initial load. Get and show reviewList from the db if any
+            val reviewsInDb = reviewListRepository.getCachedProductReviews()
+                .filter { it.product?.remoteProductId == navArgs.remoteProductId }
```
@JorgeMucientes is there a reason we are filtering on the code here instead of using `getProductReviewsForProductAndSiteId` that would filter on the DB itself? Unless I'm missing something, we can improve the ReviewsListRepostory with a function that uses this to avoid loading all reviews, I think that loading the whole reviews list from the DB just to iterate over them and filter by the product ID will have a noticeable impact for big stores.

JorgeMucientes replied:
Hmm, not really, good point! I've refactored `getCachedProductReviews` function to support filtering by `productId` when needed: https://github.com/woocommerce/woocommerce-android/pull/9753/commits/48cc7372e20c0617b816e17dd19c2c6b0c7f8c01

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/reviews/ProductReviewsViewModel.kt:
```diff
@@ -87,101 +96,118 @@ class ProductReviewsViewModel @Inject constructor(
         }
+    }
 
+    private suspend fun applyUnreadFilter(loadMore: Boolean) {
```
np, the name `applyUnreadFilter` is a bit misleading, as now this fetches the reviews and not just applies a filter, WDYT?

JorgeMucientes replied:
Agreed! Renamed it https://github.com/woocommerce/woocommerce-android/pull/9753/commits/5a08b67cf939f8f833a0e12425373fa68c52051f

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/reviews/ProductReviewsViewModel.kt:
```diff
@@ -87,101 +96,118 @@ class ProductReviewsViewModel @Inject constructor(
         )
     }
 
+    private fun loadMoreUnreadIfNeeded(unreadReviews: List<ProductReview>) {
+        if (unreadReviews.size < NUMBER_OF_ITEMS_TO_MAKE_CONTENT_SCROLLABLE) {
+            productReviewsViewState = productReviewsViewState.copy(isLoadingMore = true)
+            fetchProductReviews(navArgs.remoteProductId, loadMore = true)
+        }
+    }
```
I think this will have a similar issue to what we are trying to avoid with the new changes, and we can easily avoid it: the `NotificationModel` holds the `id` of the product too (the id `post` in the `Ids` object), so if we improve the function `fetchOnlyUnreadProductReviews` by adding an additional argument for the product ID (or create a new overload), we can then filter the unread reviews by the product ID too, which would give us a deterministic result, and we won't need this logic for additional pages here.

JorgeMucientes replied:
> I think this will have a similar issue to what we are trying to avoid with the new changes

Not really, keep in mind that now we are only fetching `unread` reviews. And the number of unread reviews is limited by the notifications we fetch from the API. If I'm not missing something we fetch 200 notification maximum judging by the value of `NOTIFICATION_DEFAULT_NUMBER` in `NotificationRestClient`. That means that in the worst case scenario we'd be fetching 200 unread reviews in chunks of 25. 

Anyway, all that doesn't matter anymore. The reason I opted for that approach was because I was unaware of "NotificationModel holds the id of the product too (the id post in the Ids object)". Given that, code can be simpler and avoid all this `loadMoreIfNeeded` workaround. Great insight, thanks! https://github.com/woocommerce/woocommerce-android/pull/9753/commits/66670ec13e64f3c9e555a64249e1ed40d4a8578f

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/reviews/ProductReviewsViewModel.kt:
```diff
@@ -87,101 +96,118 @@ class ProductReviewsViewModel @Inject constructor(
-        if (result.isError) {
-            AnalyticsTracker.track(
-                PRODUCT_REVIEWS_LOAD_FAILED,
-                mapOf(
-                    AnalyticsTracker.KEY_ERROR_CONTEXT to this::class.java.simpleName,
-                    AnalyticsTracker.KEY_ERROR_TYPE to result.error?.type?.toString(),
-                    AnalyticsTracker.KEY_ERROR_DESC to result.error?.message
-                )
```
Are we OK with losing these properties 🤔 ? I know it's because of the refactor and having an enum as a result, but we can easily make it a sealed interface, and then update the `Error` state to include details about the error, WDYT?

JorgeMucientes replied:
Tbh, I'd bet no one is looking at these properties and most likely ever would 😅 . And honestly, given the time constraints I had for the task I didn't want to invest more time in something like this. But is true that is not my call to make. So I'll fix it to avoid loosing info here. 
I think making it a sealed interface won't be as straightforward as it looks as there are tons of usages `RequestResult` enum, so the refactor would be massive. I'll create a new return type specifically for `ReviewListRepository`

hichamboushaba replied:
If you think it's OK to remove them, just ignore the comment, it's not a blocker 🙂.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListRepository.kt:
```diff
@@ -195,6 +199,51 @@ class ReviewListRepository @Inject constructor(
+            site = selectedSite.get(),
+            filterBySubtype = listOf(STORE_REVIEW.toString())
+        )
+            .filter { !it.read }
```
The `read` property is a column, so we can easily avoid the filtering here by retrieving only the `unread` notifications, and considering that sites could have a big number of reviews, this will be a significant improvement.

JorgeMucientes replied:
Hmm, I'm not sure this optimization would have any impact at all. And it would require several FluxC changes. As mentioned earlier it seems we only fetch 200 notifications by default. So worst-case scenario we would be operating with a list of 200 elements. With this order of magnitude, I don't think it makes much difference in filtering or ordering in the SQL query or here.

hichamboushaba replied:
Sounds good, it's not a blocker. I didn't know about the limit of 200 notifications, so it should be OK as you said.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListRepository.kt:
```diff
@@ -195,6 +199,51 @@ class ReviewListRepository @Inject constructor(
+                    reviewIds = unreadProductReviewIdsToFetch,
+                    offset = 0 // Must be zero so the API filters only by ids and not page offset
+                ),
+                deletePreviouslyCachedReviews = !loadMore
```
In FluxC, you mentioned that for unread reviews, we won't be deleting cached reviews, and I agree with this, can you check if this is just an accidental error here?

JorgeMucientes replied:
Hmm, I honestly can't remember why I changed to this in the end. But after re-testing I don't see a reason for not having this set to false always: https://github.com/woocommerce/woocommerce-android/pull/9753/commits/d6c74a4c93cc08a0943028ad396f964879f3da37

hichamboushaba reviewed (commented):
Nice work @JorgeMucientes 👏, the implementation is nice, and I'm sure we are on the right track, we just need to smooth some rough edges.

I left some comments regarding the product reviews screen.
And for the reviews list, I found one issue, as you can see in the following video, the sorting is not deterministic, and thus causes the reviews to jump when loading next pages:

https://github.com/woocommerce/woocommerce-android/assets/1657201/9248ca71-2b93-4235-be7a-0206ae9b221f

I believe the cause is that reviews and their notifications don't necessarily have the same timestamp, and since we use the date for sorting, this causes this jumping issue.
I suggest we use the `remoteProductReviewId` for sorting, if we sort it in a descending way, it should still have the same effect as the `date_created`, since WP would increase it for new reviews, (I didn't confirm this theory, but anyway, even if we don't sort the reviews in a date descending way when using the `unread` filter, I think it should still be OK). Something to **keep in mind**, if we agree on the sorting, then the same sorting should be used for both the notifications and the reviews, let me know what you think 🙂.

JorgeMucientes committed:
Add support to get cached product reviews by product id

JorgeMucientes committed:
Support fetching unread reviews only for provided product Id

JorgeMucientes committed:
Minor function renaming to improve readability

JorgeMucientes committed:
Order unread reviews by desc chronological order of remoteId value

JorgeMucientes committed:
Remove unused constant

JorgeMucientes commented:
Hey @hichamboushaba thanks for the careful review and great feedback :kudos: 🏅 

I'll address most of the suggested changes and reply individually to your comments :) 

About this: 
> I believe the cause is that reviews and their notifications don't necessarily have the same timestamp, and since we use the date for sorting, this causes this jumping issue.
I suggest we use the remoteProductReviewId for sorting, if we sort it in a descending way, it should still have the same effect as the date_created, since WP would increase it for new reviews, (I didn't confirm this theory, but anyway, even if we don't sort the reviews in a date descending way when using the unread filter, I think it should still be OK). Something to keep in mind, if we agree on the sorting, then the same sorting should be used for both the notifications and the reviews, let me know what you think 🙂.

Yeah I noticed it too while testing. But I believe it is because we are using the Plugin "Virtual Reviews" to generate bulks of fake reviews with the same timestamp. Or even create product reviews with a timestamp in the past. But if you create real reviews manually this is super unlikely to happen, that is why I didn't pay much attention to it. 

I think your suggestion of sorting by `unreadProductReviewIds` field by `remoteReviewId` is a great idea and should work, as long as we don't use plugins like the one we used for testing that enable you to create reviews for past dates. I've checked and review Ids always follow chronological order: 
```
 Product reviewId: 234  NOTIFICATION TIMESTAMP: 2023-09-15T15:30:11+00:00
 Product reviewId: 233  NOTIFICATION TIMESTAMP: 2023-09-15T13:11:01+00:00
 Product reviewId: 232  NOTIFICATION TIMESTAMP: 2023-09-15T09:14:59+00:00
 Product reviewId: 231  NOTIFICATION TIMESTAMP: 2023-09-15T09:14:42+00:00
 Product reviewId: 229  NOTIFICATION TIMESTAMP: 2023-09-05T15:19:06+00:00
 Product reviewId: 228  NOTIFICATION TIMESTAMP: 2023-09-05T15:19:06+00:00
 Product reviewId: 227  NOTIFICATION TIMESTAMP: 2023-09-05T15:19:06+00:00
 Product reviewId: 226  NOTIFICATION TIMESTAMP: 2023-09-05T15:19:06+00:00
 Product reviewId: 225  NOTIFICATION TIMESTAMP: 2023-09-05T15:19:06+00:00
 Product reviewId: 224  NOTIFICATION TIMESTAMP: 2023-09-05T15:19:06+00:00
 Product reviewId: 223  NOTIFICATION TIMESTAMP: 2023-09-05T15:19:06+00:00
 Product reviewId: 222  NOTIFICATION TIMESTAMP: 2023-09-05T15:19:05+00:00
 Product reviewId: 221  NOTIFICATION TIMESTAMP: 2023-09-05T15:19:05+00:00
 Product reviewId: 220  NOTIFICATION TIMESTAMP: 2023-09-05T15:19:05+00:00
 Product reviewId: 219  NOTIFICATION TIMESTAMP: 2023-09-05T15:19:05+00:00
...
```

However, we can rely on this type of sorting by `remoteReviewId` only for unread reviews that have a matching notification, (which is actually where we need it). But when fetching product reviews directly from the API, the `reviewIds` are not returned in chronological order, so we need to still rely on the `timeStamp` value. 

```
ReviewId: 233 created: Fri Sep 15 22:10:59 GMT+09:00 2023 notificationDate: 2023-09-15T13:11:01+00:00
ReviewId: 232 created: Fri Sep 15 18:14:59 GMT+09:00 2023 notificationDate: 2023-09-15T09:14:59+00:00
ReviewId: 231 created: Fri Sep 15 18:14:41 GMT+09:00 2023 notificationDate: 2023-09-15T09:14:42+00:00
ReviewId: 204 created: Wed Sep 06 08:00:00 GMT+09:00 2023 notificationDate: 2023-09-05T15:18:43+00:00
ReviewId: 134 created: Wed Sep 06 08:00:00 GMT+09:00 2023 notificationDate: null
ReviewId: 109 created: Wed Sep 06 08:00:00 GMT+09:00 2023 notificationDate: null
ReviewId: 42 created: Wed Sep 06 08:00:00 GMT+09:00 2023 notificationDate: null
ReviewId: 229 created: Wed Sep 06 07:00:00 GMT+09:00 2023 notificationDate: 2023-09-05T15:19:06+00:00
ReviewId: 214 created: Wed Sep 06 07:00:00 GMT+09:00 2023 notificationDate: 2023-09-05T15:18:45+00:00
ReviewId: 203 created: Wed Sep 06 07:00:00 GMT+09:00 2023 notificationDate: 2023-09-05T15:18:43+00:00
ReviewId: 194 created: Wed Sep 06 07:00:00 GMT+09:00 2023 notificationDate: 2023-09-05T15:18:42+00:00
ReviewId: 182 created: Wed Sep 06 07:00:00 GMT+09:00 2023 notificationDate: 2023-09-05T15:18:26+00:00
ReviewId: 119 created: Wed Sep 06 07:00:00 GMT+09:00 2023 notificationDate: null
ReviewId: 202 created: Wed Sep 06 06:00:00 GMT+09:00 2023 notificationDate: 2023-09-05T15:18:43+00:00
ReviewId: 193 created: Wed Sep 06 06:00:00 GMT+09:00 2023 notificationDate: 2023-09-05T15:18:42+00:00
```
Anyway I'll apply the fix for unread reviews to sort them by `reviewId` values.

hichamboushaba commented:
> However, we can rely on this type of sorting by remoteReviewId only for unread reviews that have a matching notification, (which is actually where we need it). But when fetching product reviews directly from the API, the reviewIds are not returned in chronological order, so we need to still rely on the timeStamp value.

Yes, this was my idea too 👍

JorgeMucientes committed:
Avoid removing cached unread reviews when enabling filter

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListRepository.kt:
```diff
@@ -195,6 +199,55 @@ class ReviewListRepository @Inject constructor(
+            site = selectedSite.get(),
+            filterBySubtype = listOf(STORE_REVIEW.toString())
+        )
+            .filter { !it.read }
+            .filter { if (productId == null) true else it.meta?.ids?.post == productId }
```
np, combining the checks in a single `filter` would be more optimized, two consecutive `filter` calls would cause two iterations over the list.

JorgeMucientes replied:
True! Updated it, merged both filtering conditions. 
However, one principle I try to follow in general is "code readability over performance optimizations, presuming the impact on performance from UX perspective is unnoticeable. In this case, I haven't measured it and the change doesn't make the code much less readable so I think it is fine to merge the conditions.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListRepository.kt:
```diff
@@ -195,6 +199,55 @@ class ReviewListRepository @Inject constructor(
+            .filter { !it.read }
+            .filter { if (productId == null) true else it.meta?.ids?.post == productId }
+            .map { it.getCommentId() }
+            .sortedByDescending { it }
```
Future improvement: handle the sorting in SQL 🙂.

hichamboushaba reviewed (approved):
Excellent work @JorgeMucientes 🎖️, and really sorry for the long review.
This is ready for merge once the FluxC PR is merged and the commit is updated.

JorgeMucientes committed:
Avoid extra list iteration by filtering once

JorgeMucientes committed:
Update FluxC changeset

JorgeMucientes committed:
Merge branch 'trunk' into hack-week/unread-product-reviews-filter

# Conflicts:
#	build.gradle

JorgeMucientes committed:
Fix release notes formatting
